external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -1726,6 +1795,33 @@ fn loop_body(
     }
 }
 
+fn exp_to_ref(context: &mut Context, loc: Loc, mut_: bool, e: Box<T::Exp>) -> Box<T::Exp> {
+    let ety = &e.ty;
+    let current_ty = core::unfold_type(&context.subst, ety.clone());
+    match current_ty.value {
+        Type_::Ref(false, _t) => {
+            if mut_ {
+                make_error_exp(context, loc)
+            } else {
+                e
+            }
+        }
+        Type_::Ref(true, t) => {
+            if mut_ {
+                e
+            } else {
+                let freeze_type = sp(loc, Type_::Ref(false, Box::new(*t)));
+                let freeze_exp = sp(
+                    loc,
+                    T::UnannotatedExp_::Annotate(e, Box::new(freeze_type.clone())),
+                );
```

Comment:
Does the HLIR not automatically do this?

Reply:
This annotation is what causes it to do this.